src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3760,93 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
+            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, TypeSymbol receiverType, WellKnownMember wellKnownMember)
             {
-                var wellKnownMethod = compilation.GetWellKnownTypeMember(wellKnownMember);
-                if (wellKnownMethod is null)
+                var wellKnownMethod = (MethodSymbol)compilation.GetWellKnownTypeMember(wellKnownMember);
+                if (wellKnownMethod is null || receiverType is null)
                 {
                     return false;
                 }
 
                 var wellKnownType = wellKnownMethod.ContainingType;
                 var parameterType = method.Parameters[0].TypeWithAnnotations;
                 var constructedType = wellKnownType.Construct(ImmutableArray.Create(parameterType));
+                var constructedMethod = wellKnownMethod.AsMember(constructedType);
 
-                Symbol constructedMethod = null;
-                foreach (var member in constructedType.GetMembers(WellKnownMemberNames.ObjectEquals))
+                // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
+                if (constructedMethod.Equals(method))
                 {
-                    if (member.OriginalDefinition.Equals(wellKnownMethod))
-                    {
-                        constructedMethod = member;
-                        break;
-                    }
+                    return true;
                 }
 
-                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
+                // we won't be able to find an implicit implementation of 'constructedMethod' on a derived interface.
+                if (method.ContainingType.IsInterface)
+                {
+                    return false;
+                }
 
-                // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
-                if (constructedMethod.Equals(method))
+                // check whether 'method', when called on this receiver, is an implementation of 'constructedMethod'.
+                for (var baseType = receiverType; baseType is object && method is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
                 {
-                    return true;
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
+                    {
+                        // we know no base type will implement this interface member either
+                        return false;
+                    }
+
+                    // 'method' definitely won't match if the implementation is explict.
```

Comment:
>/ 'method' definitely won't match if the implementation is explict. [](start = 21, length = 67)

I do not think this is correct. It is definitely possible to get a callable method defined in metadata that implements interface explicitly. That method, BTW, can explicitly implement even an unrelated interface. Please, add a tests for both situations. I think there was no problem with the ```if (implementationMethod.ContainingType.IsInterface)``` check, you just needed to add a test.

Reply:
In order to test this I need to define the implementing class in IL right? Are there existing tests I should look at as a starting point for this?

Reply:
> In order to test this I need to define the implementing class in IL right? Are there existing tests I should look at as a starting point for this?

- You can define a type in VB, all interface implementations in VB are explicit. 
- You can decompile a C# type with explicit implementation and remove the interface name prefix form method's name and make it public

---
In reply to: [390628812](https://github.com/dotnet/roslyn/pull/41918#discussion_r390628812) [](ancestors = 390628812)

Reply:
We have a helper CreateCompilationWithIL to "include" IL into a compilation

---
In reply to: [390630378](https://github.com/dotnet/roslyn/pull/41918#discussion_r390630378) [](ancestors = 390630378,390628812)

Reply:
I think I've added the suggested tests for callable explicit implementations. But it feels like I'm still missing something.

Reply:
> I think I've added the suggested tests for callable explicit implementations. But it feels like I'm still missing something.

Still have this feeling?

---
In reply to: [390702059](https://github.com/dotnet/roslyn/pull/41918#discussion_r390702059) [](ancestors = 390702059)

Reply:
No, now that I've figured out what scenario actually makes us find an 'implementationMethod' defined in an interface, I feel more confident.